src/Microsoft.Data.Analysis.IO/DataFrameIO.cs

```diff
@@ -0,0 +1,266 @@
+                                Encoding encoding = null)
+        {
+            Stream stream = csvStream();
+            if (!stream.CanSeek)
+            {
+                throw new ArgumentException(Strings.NonSeekableStream, nameof(stream));
+            }
+
+            if (dataTypes == null && guessRows <= 0)
+            {
+                throw new ArgumentException(string.Format(Strings.ExpectedEitherGuessRowsOrDataTypes, nameof(guessRows), nameof(dataTypes)));
+            }
+
+            var linesForGuessType = new List<string[]>();
+            long rowline = 0;
+            int numberOfColumns = dataTypes?.Length ?? 0;
+
+            if (header == true && numberOfRowsToRead != -1)
+            {
+                numberOfRowsToRead++;
+            }
+
+            List<DataFrameColumn> columns;
+            // First pass: schema and number of rows.
+            using (TextFieldParser parser = new TextFieldParser(stream, defaultEncoding: encoding ?? Encoding.UTF8))
+            {
+                parser.SetDelimiters(separator.ToString());
+                SetTextFieldParserProperties(parser);
+                while (!parser.EndOfData)
+                {
+                    if ((numberOfRowsToRead == -1) || rowline < numberOfRowsToRead)
+                    {
+                        if (linesForGuessType.Count < guessRows || (header && rowline == 0))
+                        {
+                            var spl = parser.ReadFields();
+                            if (header && rowline == 0)
+                            {
+                                if (columnNames == null)
+                                {
+                                    columnNames = spl;
+                                }
+                            }
+                            else
+                            {
+                                linesForGuessType.Add(spl);
+                                numberOfColumns = Math.Max(numberOfColumns, spl.Length);
+                            }
+                        }
+                    }
+                    ++rowline;
+                    if (rowline == guessRows || guessRows == 0)
+                    {
+                        break;
+                    }
+                }
+
+                if (rowline == 0)
+                {
+                    throw new FormatException(Strings.EmptyFile);
+                }
+
+                columns = new List<DataFrameColumn>(numberOfColumns);
+                // Guesses types or looks up dataTypes and adds columns.
+                for (int i = 0; i < numberOfColumns; ++i)
+                {
+                    Type kind = dataTypes == null ? GuessKind(i, linesForGuessType) : dataTypes[i];
+                    columns.Add(CreateColumn(kind, columnNames, i));
+                }
+            }
+
+            using (TextFieldParser parser = new TextFieldParser(csvStream(), defaultEncoding: encoding ?? Encoding.UTF8))
```

Comment:
TextFieldParser seems to be forward-only. So I'm making 2 calls here as a quick proof of concept